File: test/test-coff.mjs
import assert from "node:assert";
import fs from "node:fs";
import path from "node:path";
import test, { describe, it } from "node:test";
import url from "node:url";
import { NodeBufferReader, SubFileReader } from "../src/reader.mjs";
import {
  findCOFFSectionsByNameAsync,
  getCOFFSectionsAsync,
} from "../src/coff.mjs";

let __filename = url.fileURLToPath(import.meta.url);
let __dirname = path.dirname(__filename);

describe("sections", (t) => {
  it("small.obj findCOFFSectionsByNameAsync", async () => {
    let file = new NodeBufferReader(
      await fs.promises.readFile(path.join(__dirname, "coff/small.obj"))
    );
    let sectionReaders = await findCOFFSectionsByNameAsync(file, ".debug$S");

    // Data according to: dumpbin.exe /HEADERS
    assert.strictEqual(sectionReaders.length, 1);
    assert.ok(sectionReaders[0] instanceof SubFileReader);
    assert.strictEqual(sectionReaders[0].baseReader, file);
    assert.strictEqual(sectionReaders[0].subFileOffset, 0x10b);
    assert.strictEqual(sectionReaders[0].subFileSize, 0x1e8);
  });

  it("small.obj", async () => {
    let file = new NodeBufferReader(
      await fs.promises.readFile(path.join(__dirname, "coff/small.obj"))
    );
    let sections = await getCOFFSectionsAsync(file);
    sections = sections.map((section) => ({
      name: section.name,
      dataSize: section.dataSize,
      dataFileOffset: section.dataFileOffset,
    }));

    // Data according to: dumpbin.exe /HEADERS
    assert.deepStrictEqual(sections, [
      {
        name: ".drectve",
        dataSize: 0x2f,
        dataFileOffset: 0xdc,
      },
      {
        name: ".debug$S",
        dataSize: 0x1e8,
        dataFileOffset: 0x10b,
      },
      {
        name: ".text$mn",
        dataSize: 0x3,
        dataFileOffset: 0x31b,
      },
      {
        name: ".debug$T",
        dataSize: 0x68,
        dataFileOffset: 0x31e,
      },
      {
        name: ".chks64",
        dataSize: 0x28,
        dataFileOffset: 0x386,
      },
    ]);
  });
});

File: test/test-coff.cpp
#include <cppstacksize/example-file.h>
#include <cppstacksize/pe.h>
#include <cppstacksize/reader.h>
#include <gtest/gtest.h>

namespace cppstacksize {
namespace {
TEST(Test_COFF, find_coff_sections_by_name_in_small_obj) {
  Example_File file("coff/small.obj");
  PE_File pe = parse_pe_file(&file.reader());
  std::vector<Sub_File_Reader<Span_Reader>> section_readers =
      pe.find_sections_by_name(u8".debug$S");

  // Data according to: dumpbin.exe /HEADERS
  ASSERT_EQ(section_readers.size(), 1);
  EXPECT_EQ(section_readers[0].base_reader(), &file.reader());
  EXPECT_EQ(section_readers[0].sub_file_offset(), 0x10b);
  EXPECT_EQ(section_readers[0].size(), 0x1e8);
}

TEST(Test_COFF, small_obj_sections) {
  Example_File file("coff/small.obj");
  PE_File pe = parse_pe_file(&file.reader());

  // Data according to: dumpbin.exe /HEADERS
  ASSERT_EQ(pe.sections.size(), 5);
  EXPECT_EQ(pe.sections[0].name, u8".drectve");
  EXPECT_EQ(pe.sections[0].data_size, 0x2f);
  EXPECT_EQ(pe.sections[0].data_file_offset, 0xdc);
  EXPECT_EQ(pe.sections[1].name, u8".debug$S");
  EXPECT_EQ(pe.sections[1].data_size, 0x1e8);
  EXPECT_EQ(pe.sections[1].data_file_offset, 0x10b);
  EXPECT_EQ(pe.sections[2].name, u8".text$mn");
  EXPECT_EQ(pe.sections[2].data_size, 0x3);
  EXPECT_EQ(pe.sections[2].data_file_offset, 0x31b);
  EXPECT_EQ(pe.sections[3].name, u8".debug$T");
  EXPECT_EQ(pe.sections[3].data_size, 0x68);
  EXPECT_EQ(pe.sections[3].data_file_offset, 0x31e);
  EXPECT_EQ(pe.sections[4].name, u8".chks64");
  EXPECT_EQ(pe.sections[4].data_size, 0x28);
  EXPECT_EQ(pe.sections[4].data_file_offset, 0x386);
}
}
}
